ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Forging.hs

```diff
@@ -31,195 +30,182 @@ import           Ouroboros.Consensus.HardFork.Combinator.Ledger (Ticked (..))
-    --
-    -- * When one or more eras failed to update, we return 'UpdateFailed' with
-    --   the eras that didn't fail left out of the
-    --   'HardForkForgeStateUpdateError' ('OptNP').
-    --
-    -- * When one or more eras updated, we return 'Updated' even when some
-    --   eras didn't. The 'HardForkForgeStateInfo' ('OptNP') will contain all
-    --   eras that were included in the 'BlockForging' 'OptNP', including the
-    --   eras that returned 'Unchanged'.
-    --
-    -- * When all eras were unchanged, we return 'Unchanged'. The
-    --   'HardForkForgeStateInfo' ('OptNP') will contain all eras that were
-    --   included in the 'BlockForging' 'OptNP'.
-    undistribForgeStateUpdateInfo ::
-         OptNP 'False ForgeStateUpdateInfo xs
+    undistrib ::
+         NS ForgeStateUpdateInfo xs
       -> ForgeStateUpdateInfo (HardForkBlock xs)
-    undistribForgeStateUpdateInfo updateInfos = ForgeStateUpdateInfo $
-        case go updateInfos of
-          Updated      infos -> Updated      (PerEraForgeStateInfo        infos)
-          Unchanged    infos -> Unchanged    (PerEraForgeStateInfo        infos)
-          UpdateFailed errs  -> UpdateFailed (PerEraForgeStateUpdateError errs)
+    undistrib = hcollapse . hzipWith3 inj injections injections
       where
-        go :: SListI xs'
-           => OptNP empty ForgeStateUpdateInfo xs'
-           -> UpdateInfo (OptNP 'False WrapForgeStateInfo        xs')
-                         (OptNP empty  WrapForgeStateInfo        xs')
-                         (OptNP 'False WrapForgeStateUpdateError xs')
-        go OptNil         = Unchanged OptNil
-        go (OptCons x xs) = consUpdateInfo x (go xs)
-        go (OptSkip   xs) = skipUpdateInfo (go xs)
+        inj :: forall blk.
+               Injection WrapForgeStateInfo xs blk
+            -> Injection WrapForgeStateUpdateError xs blk
+            -> ForgeStateUpdateInfo blk
+            -> K (ForgeStateUpdateInfo (HardForkBlock xs)) blk
+        inj injInfo injUpdateError forgeStateUpdateInfo =
+            K $ ForgeStateUpdateInfo $
+              case getForgeStateUpdateInfo forgeStateUpdateInfo of
+                Updated      info -> Updated      $ injInfo'        info
+                Unchanged    info -> Unchanged    $ injInfo'        info
+                UpdateFailed err  -> UpdateFailed $ injUpdateError' err
+          where
+            injInfo' ::
+                 ForgeStateInfo blk
+              -> OneEraForgeStateInfo xs
+            injInfo' =
+                  OneEraForgeStateInfo
+                . unK
+                . apFn injInfo
+                . WrapForgeStateInfo
+
+            injUpdateError' ::
+                 ForgeStateUpdateError blk
+              -> OneEraForgeStateUpdateError xs
+            injUpdateError' =
+                  OneEraForgeStateUpdateError
+                . unK
+                . apFn injUpdateError
+                . WrapForgeStateUpdateError
 
 hardForkCheckCanForge ::
```

Comment:
Add precondition: `NS BlockForging`, ticked chain dep state, `HardForkIsLeader` and `HardForkForgeStateInfo` are all aligned. This follows from the postconditions of `check(IsLeader)` and `hardForkUpdateForgeState`.